nixos/firejail: wrappedPackages

ryneeverett opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This is a rebase and continuation of the work at #45469. See commit messages.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
nixos/firejail: add wrappedPackages option

Using this, not only the wrapped binaries but also all other
directories from the derivations are put into /run/current-system/sw.
This is useful if you need desktop entries, icons, etc.

ryneeverett committed:
nixos/firejail: update and test wrappedPackages

- Remove unnecessary buildInputs and add indentation per review of the
original PR #45469.
- Warn that desktop files won't be firejailed if they have an absolute
path per discussion in #45469.
- Update wrappedPackages to use runtimeShell and exec per #70477.
- Test that wrappedPackages works.

symphorien commented on nixos/modules/programs/firejail.nix:
```diff
@@ -78,12 +98,30 @@ in {
+        However, you should be careful about using these packages'
+        libraries as they will not be wrapped. Note also that applications may
+        not be firejailed if invoked via a desktop file that specifies an
+        absolute path to the unwrapped binary.
```
How about adding a check such that if a desktop file contains `Exec=/` then the build fails ?

symphorien replied:
Oh actually such a check exists. I'd fail the build instead of printing a warning, this way there is no "risk" for the user.

ryneeverett replied:
Agreed.

ryneeverett replied:
I just ran across a couple issues which I think solidify the position that hard coded store paths in desktop files should be regarded as bugs: #106996, #101859.

symphorien commented:
Would you be interested in adding support for custom options (at least, a custom profile)?

ryneeverett committed:
nixos/firejail: error out on absolute desktop file

Per @symphorien's review, it's better to error out when a desktop file
specifies an absolute path to the unwrapped binary than to echo a
warning which will likely be missed.

In cases where the desktop file cannot be wrapped, the same behavior can
be achieved by adding the binaries to wrappedBinaries and doing a
conventional installation of the given package.

ryneeverett commented:
> Would you be interested in adding support for custom options (at least, a custom profile)?

I'm not particularly motivated to do that but I can see where a similar backwards compatible strategy as you implemented in #103813 would work here. Maybe I don't fully understand the use case -- wouldn't profiles in ~/.config/firejail or /etc/firejail work here? (I'll admit I haven't tested it.) Isn't that sufficient for all practical desktop use cases?

symphorien commented:
If I'm not mistaken, firejail picks up a generic (permissive) profile by default. By running a program with a dedicated profile (and upstream provides many) you get a tighter sandbox.
Adding profiles in ~/config/firejail does not really help as it only registers new profiles which are not used by default.
My use case for additional options is to use `--private=/some/dir` so that some proprietary programs I must run cannot see my home directory.

ryneeverett committed:
nixos/firejail: canonicalize wrapped binary path

Many packages have symlinked aliases which don't work when just passing
the binary name to firejail, presumably because the binary name points
to the firejailed wrapper which results in the firejailed wrapper
calling another firejailed wrapper.

For example, before this change the libreoffice wrapper called "firejail
soffice" where the soffice wrapper called "firejail /nix/store...soffice".
The soffice wrapper worked but the libreoffice wrapper yielded:

    firejail: /nix/store/v6l2sacryfr88yqq0pq7sia8wfgm9q31-wrapper.c:203: main: Assertion `!(st.st_mode & S_ISUID) || (st.st_uid == geteuid())' failed.

ryneeverett commented:
> If I'm not mistaken, firejail picks up a generic (permissive) profile by default.

Fortunately you seem to be mistaken. Testing this out with libreoffice, firejail does the right thing by reading the upstream profiles:

```console
$ libreoffice
Reading profile /nix/store/1ry4v2vnyli2gpi5h0cx2sm8bmjji4yv-firejail-0.9.62/etc/firejail/soffice.profile
Reading profile /nix/store/1ry4v2vnyli2gpi5h0cx2sm8bmjji4yv-firejail-0.9.62/etc/firejail/libreoffice.profile
Reading profile /nix/store/1ry4v2vnyli2gpi5h0cx2sm8bmjji4yv-firejail-0.9.62/etc/firejail/libreoffice.local
...
```

> Adding profiles in ~/config/firejail does not really help as it only registers new profiles which are not used by default.
> My use case for additional options is to use --private=/some/dir so that some proprietary programs I must run cannot see my home directory.

I'm not completely familiar with how firejail's profile precedence works, but local overrides are supposed to override the default profiles. If they do not that is a bug in the nix implementation which it would be better to fix if possible rather than create configuration options to work around it. My best understanding is that local overrides currently do work: https://github.com/NixOS/nixpkgs/pull/83515.

symphorien commented:
> Fortunately you seem to be mistaken.

Oh I had only tested with zoom, for which it does not work. It looks like it is due to the fact that the executable is called zoom-us and the profile zoom.

symphorien reviewed (changes requested):
I tried this with gimp and the wrapping is circular


![image](https://user-images.githubusercontent.com/12595971/106364022-f4fa9480-6323-11eb-88b1-a2d61c016e6e.png)

ryneeverett commented:
> I tried this with gimp and the wrapping is circular

I can't reproduce this on stable:

```console
$ head $(which gimp)
#! /nix/store/a3fc4zqaiak11jks9zd579mz5v0li8bg-bash-4.4-p23/bin/bash -e
exec /run/wrappers/bin/firejail "/nix/store/4a0d7inzfq8vc09lqbmp5px9n6czvs7y-gimp-2.10.20/bin/gimp-2.10" "$@"
```

Any chance you have gimp firejailed elsewhere such as in wrappedBinaries or an override?

symphorien commented:
Nope that was in a nixos vm with full config:
```
{ config, pkgs, lib, ...}: {
  config = {
    users.users.root.initialPassword = "root";
    users.users.alice = {
      isNormalUser = true;
      initialPassword = "alice";
    };
    console = {
      font = "Lat2-Terminus16";
      keyMap = "fr";
    };
    programs.firejail = {
      enable = true;
      wrappedPackages = [
        pkgs.gimp
      ];
    };
    services.xserver = {
      enable = true;
      layout = "fr";
      xkbVariant = "oss";
      libinput = {
        enable = true;
        tapping = false;
      };
      xkbOptions = "caps:swapescape, compose:prsc";
      displayManager = {
        lightdm.enable = true;
        defaultSession = "xfce";
      };
      desktopManager = {
        xterm.enable = false;
        xfce = {
          enable = true;
        };
      };
    };
  };
}
```

ryneeverett commented:
First off thanks for introducing me to nixos vm's! Somehow I'd never noticed the build-vm subcommand before.

I still can't reproduce the circular wrapping though and made only the necessary changes so that I could use the keyboard:

```nix
{ config, pkgs, lib, ...}: {
  config = {
    users.users.root.initialPassword = "root";
    users.users.alice = {
      isNormalUser = true;
      initialPassword = "alice";
    };
    console = {
      font = "Lat2-Terminus16";
      # keyMap = "fr";
    };
    programs.firejail = {
      enable = true;
      wrappedPackages = [
        pkgs.gimp
      ];
    };
    services.xserver = {
      enable = true;
      # layout = "fr";
      # xkbVariant = "oss";
      libinput = {
        enable = true;
        tapping = false;
      };
      xkbOptions = "caps:swapescape, compose:prsc";
      displayManager = {
        lightdm.enable = true;
        defaultSession = "xfce";
      };
      desktopManager = {
        xterm.enable = false;
        xfce = {
          enable = true;
        };
      };
    };
  };
}

```

![image](https://user-images.githubusercontent.com/3280280/106371551-b53ba900-6333-11eb-9c38-48a66e99592f.png)

I don't know which environment variable is bothering firejail in this system but it doesn't seem to be caused by this patch.

symphorien commented:
For full reproducibility:
on top of 03c4f7d7dff27a93a47d8b013ccb71e0b1223228 (your PR), and with this english keymap configuration:
```
{ config, pkgs, lib, ...}: {
  config = {
    users.users.root.initialPassword = "root";
    users.users.alice = {
      isNormalUser = true;
      initialPassword = "alice";
    };
    console = {
      font = "Lat2-Terminus16";
      # keyMap = "fr";
    };
    programs.firejail = {
      enable = true;
      wrappedPackages = [
        pkgs.gimp
      ];
    };
    services.xserver = {
      enable = true;
      # layout = "fr";
      # xkbVariant = "oss";
      libinput = {
        enable = true;
        tapping = false;
      };
      # xkbOptions = "caps:swapescape, compose:prsc";
      displayManager = {
        lightdm.enable = true;
        defaultSession = "xfce";
      };
      desktopManager = {
        xterm.enable = false;
        xfce = {
          enable = true;
        };
      };
    };
  };
}
```
then when in run ` nixos-rebuild build-vm -I nixpkgs=$(pwd) -I nixos-config=$(readlink -f vm.nix)` I get 
`/nix/store/him8q86pfk4yf5if00hjk1dqpkycs9nl-nixos-vm/bin/run-nixos-vm`
After removing `nixos.qcow2` i can still reproduce the issue.

![image](https://user-images.githubusercontent.com/12595971/106387598-5e38e100-63d2-11eb-8a77-8e466fc88ef1.png)


Can you check that you get the same store paths ? This should help you reproducing.

about the error message about environment variable size: I suppose firejail appends things to environment variable, so if it invokes itself, the environment variable size grows infinitely.

ryneeverett commented:
> then when in run ` nixos-rebuild build-vm -I nixpkgs=$(pwd) -I nixos-config=$(readlink -f vm.nix)` I get
> `/nix/store/him8q86pfk4yf5if00hjk1dqpkycs9nl-nixos-vm/bin/run-nixos-vm`

Hmm, with the updated config and the exact same commit and invocation I get `/nix/store/l3hn02rdsn4fj4mvwnpxiv5sp3v1fmfd-nixos-vm/bin/run-nixos-vm`.

> about the error message about environment variable size: I suppose firejail appends things to environment variable, so if it invokes itself, the environment variable size grows infinitely.

I don't see how the error message could be related to the circular wrapping issue because as you can see from my vm session above, I got this error even when the circular wrapping wasn't occurring.

ryneeverett commented:
> Hmm, with the updated config and the exact same commit and invocation I get /nix/store/l3hn02rdsn4fj4mvwnpxiv5sp3v1fmfd-nixos-vm/bin/run-nixos-vm

Just reproduced this output on a freshly provisioned machine, though I'll grant it's configuration is fairly similar to the first.

ryneeverett commented:
I pushed an additional commit using makeWrapper to simplify the implementation. I would think this would reduce the likelihood of a circular wrapping issue.

SuperSandro2000 commented on nixos/modules/programs/firejail.nix:
```diff
@@ -78,12 +98,28 @@ in {
     security.wrappers.firejail.source = "${lib.getBin pkgs.firejail}/bin/firejail";
 
-    environment.systemPackages = [ pkgs.firejail ] ++ [ wrappedBins ];
+    environment.systemPackages = [ pkgs.firejail ] ++ [ wrappedBins ] ++ wrappedPkgs;
```
```suggestion
    environment.systemPackages = [ pkgs.firejail wrappedBins ] ++ wrappedPkgs;
```

ryneeverett committed:
nixos/firejail: More compact list construction

Per SuperSandro2000's review.

ryneeverett committed:
nixos/firejail: Don't hardcode wrapperDir path

ryneeverett committed:
nixos/firejail: Attempt to remove absolute paths

Given that no consensus was reached on #68035, it's probably best to try
to patch these than to fix upstream at this point.

ryneeverett committed:
nixos/firejail: DocBook -> CommonMark
